cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -185,61 +178,68 @@ public String call(DatasetContext context) throws Exception {
      * @return sleep time in milliseconds before next fetch
      */
     private long processNotifications() {
-      boolean emptyFetch = false;
       try {
         final MessageFetcher fetcher = messagingContext.getMessageFetcher();
-        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+        String lastFetchedMessageId = Transactionals.execute(transactional, new TxCallable<String>() {
           @Override
-          public Boolean call(DatasetContext context) throws Exception {
+          public String call(DatasetContext context) throws Exception {
             return fetchAndProcessNotifications(context, fetcher);
           }
-        });
+        }, ServiceUnavailableException.class, TopicNotFoundException.class);
         failureCount = 0;
+
+        // The job queue and the last fetched message id was persisted successfully, update the local field as well.
+        // Sleep for 2 seconds if there's no notification, otherwise don't sleep
+        if (lastFetchedMessageId != null) {
+          messageId = lastFetchedMessageId;
+          return cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+        }
+        return 0L;
+
+      } catch (ServiceUnavailableException e) {
+        SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+        failureCount++;
```

Comment:
can just do `failureCount++` once outside of the try-catch block since try block will return if no exception occurs